ARCHITECTURE.md

```diff
@@ -0,0 +1,255 @@
+# Architecture
+
+## Repository layout
+
+[human-edited docs]: https://github.com/kubernetes-sigs/cli-experimental/tree/master/site
+[generated docs]: https://github.com/kubernetes-sigs/cli-experimental/tree/master/docs
+[rendered docs]: https://kubectl.docs.kubernetes.io
+[openapi]: https://kubernetes.io/blog/2016/12/kubernetes-supports-openapi
+
+[`api` module]: https://github.com/kubernetes-sigs/kustomize/blob/master/api/go.mod
+[`api`]: #the-api-module
+[`cmd/config` module]: https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/go.mod
+[`cmd/config`]: #the-cmdconfig-module
+[`kustomize` module]: https://github.com/kubernetes-sigs/kustomize/blob/master/kustomize/go.mod
+[`kustomize`]: #the-kustomize-module
+[`kyaml` module]: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/go.mod
+[`kyaml`]: #the-kyaml-module
+
+[`kyaml/kio.Filter`]: https://github.com/Kubernetes-sigs/kustomize/blob/master/kyaml/kio/kio.go
+
+[3922]: https://github.com/kubernetes-sigs/kustomize/issues/3922
+
+
+ |  directory  |             |
+ | ---------:  | :---------- |
+ |       `api` | The [`api`] module, holding high level kustomize code, suitable for import by other programs.  |
+ |       `cmd` | Various Go programs aiding repo management (see also `hack`). As an outlier, includes the special [`cmd/config`] module. |
+ |      `docs` | Old home of documentation.  New homes: [human-edited docs], [generated docs] and [rendered docs]. |
+ |  `examples` | Full kustomization examples that run as pre-commit-merge tests.  |
+ | `functions` | Examples of plugins in KRM function form. TODO([3922]): Move under `plugin`. |
+ |      `hack` | Various shell scripts to help with code mgmt. |
+ | `kustomize` | The [`kustomize`] module, holding the `main.go` for kustomize. |
+ |     `kyaml` | The [`kyaml`] module, holding Kubernetes-specific YAML editing packages used by the [`api`] module. |
+ |    `plugin` | Examples of Kustomize plugins.  |
+ | `releasing` | Instructions for releasing the various modules.  |
+ |      `site` | Old generated documentation, kept to provide redirection links to the new docs. |
+
+
+## Modules
+
+[semantically versioned]: https://semver.org
+[Go modules]: https://github.com/golang/go/wiki/Modules
+
+The [Go modules] in the kustomize repository are [semantically versioned].
```

Comment:
It could be helpful to mention any guidance around intra-repo module dependencies. E.g. it came up recently that having two mutually dependent modules, though possible if triggered by different packages, would throw a wrench in the release process.

Reply:
The first line in each module section specifies the intra-module dependence.  
I added a bullet and some italics to spring them out more.

Reply:
Yes, I see that, and it's great, but what I was hoping for was something more prescriptive than descriptive. i.e. kyaml not only doesn't have internal deps, but MUST NOT have any internal deps, and why.

Incidentally, more guidance on external dependencies would be great. E.g. be SUPER hesitant to add new ones, check licences, and absolutely never add X because of Y aspect of the kubectl integration.